go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,372 @@
+
+	CreateTableRegexp = "CREATE TABLE .* _vt.*"
+	AlterTableRegexp  = "ALTER TABLE _vt.*"
+)
+
+// all tables needed in _vt sidecar have their schema in the vtschema subdirectory
+//
+//go:embed vtschema/*
+var schemaLocation embed.FS
+
+type vtTable struct {
+	module string // which module uses this table
+	path   string // path of the schema relative to this module
+	name   string // table name
+	schema string // create table dml
+}
+
+func (t *vtTable) String() string {
+	return fmt.Sprintf("_vt.%s (%s)", t.name, t.module)
+}
+
+var vtTables []*vtTable
+var ddlCount *stats.Counter
+
+func init() {
+	initSchemaFiles()
+	ddlCount = stats.NewCounter("SidecarDbDDLQueryCount", "Number of create/upgrade queries executed")
+}
+
+func initSchemaFiles() {
+	err := fs.WalkDir(schemaLocation, ".", func(path string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !entry.IsDir() {
+			module, name := filepath.Split(path)
+			arr := strings.Split(module, "/")
+			if len(arr) >= 1 {
+				module = arr[1]
+			}
+			name = strings.Split(name, ".")[0]
+			schema, err := schemaLocation.ReadFile(path)
+			if err != nil {
+				panic(err)
+			}
+			vtTables = append(vtTables, &vtTable{name: name, module: module, path: path, schema: string(schema)})
+		}
+		return nil
+	})
+	if err != nil {
+		log.Errorf("error loading schema files: %+v", err)
+	}
+}
+
+// printCallerDetails is a helper for dev debugging
+func printCallerDetails() {
+	pc, _, line, ok := runtime.Caller(2)
+	details := runtime.FuncForPC(pc)
+	if ok && details != nil {
+		log.Infof("_vt schema init called from %s:%d\n", details.Name(), line)
+	}
+}
+
+type vtSchemaInit struct {
+	ctx            context.Context
+	exec           Exec
+	existingTables map[string]bool
+	dbCreated      bool // the first upgrade/create query will be prefixed by a _vt creation script
+}
+
+// Exec is the prototype of a callback that has to be passed to Init() to execute the specified query in the database
+type Exec func(ctx context.Context, query string, maxRows int, wantFields bool, useVT bool) (*sqltypes.Result, error)
+
+// GetDDLCount metric returns the count of sidecardb ddls that have been run as part of this vttablet's init process.
+func GetDDLCount() int64 {
+	return ddlCount.Get()
+}
+
+// Init creates or upgrades the _vt sidecar database based on declarative schema for all _vt tables
+func Init(ctx context.Context, exec Exec) error {
+	printCallerDetails() // for debug purposes only, remove in v18
+	log.Infof("Starting sidecardb Init()")
+	si := &vtSchemaInit{
+		ctx:  ctx,
+		exec: exec,
+	}
+
+	dbExists, err := si.doesVTDatabaseExist()
+	if err != nil {
+		return err
+	}
+	if !dbExists {
+		if err := si.createVTDatabase(); err != nil {
+			return err
+		}
+		si.dbCreated = true
+	}
+
+	currentDatabase, err := si.setCurrentDatabase("_vt")
+	if err != nil {
+		return err
+	}
+	// nolint
+	defer si.setCurrentDatabase(currentDatabase)
```

Comment:
I don't think we need this either. I checked that the connection used for execution is actually a new connection.